### PR TITLE
Fix recreation of console jobs on delete

### PR DESCRIPTION
**Pull request by dyson**
commit 4e16a56bc33ee626fc7637ffd64810e9ba01ecfa

    Fix console job name truncate length

    Jobs names are restricted to 63 characters as they are added as a label
    to the pods they create (and labels have a 63 character length
    restriction).

    We incorrectly limited job names thinking that the pods they created had
    a name limit of 63 characters. This fix gets us a few characters back in
    the job name and ensures our labels are the correct length.

commit e0752e6d460aa670eb9b38625d3b90688a5b7893

    Only create jobs on console creation and mark a console as destroyed if
    a console job has been deleted from underneath the console.

    Before this fix, deleting a job (even if the console had completed and
    was waiting for GC) would result in a new job being created for the
    console and therefore a new console pod executing.

    The recreation of the job is unexpected behaviour and this change
    ensures this double running cannot happen.

### CODE REVIEW

File: pkg/workloads/console/controller.go [hunk 1]
```diff
@@ -182,7 +182,14 @@ func (r *reconciler) Reconcile() (res reconcile.Result, err error) {
 		}
 	}
 
-	// Create or update the job
+	// Only create a job when the console is created. If the job is removed,
+	// return a reconciliation error.
+	if !r.console.Creating() {
+		if _, err = r.getJob(); err != nil {
+			return res, errors.Wrap(err, "failed to retrieve console job")
```
Comment (benwh) [inline pkg/workloads/console/controller.go:189]: It feels a bit odd to be returning a reconciliation error here, rather than continuing but logging a warning event, or updating the console's status (phase `Destroyed`?)

I think this may have a negative side-effect of firing alerts, because we have a metric for reconcile failures.

Comment (dyson) [reply pkg/workloads/console/controller.go:189]: Yeah I was 50/50 on this and wondering what people might think. 

My reason for the error is that deleting the job out from under the console is unexpected and shouldn't happen under normal operation. If we accept that statement as true, then I think we should error, fire alerts, be painful and attention seeking, when this happens. 

But I can see the flip side, where we say deleting a job is a valid way to kill a console, in which case we should mark the console as stopped (or more accurately like you mention, introduce a new phase called destroyed).

Having written this out now I think I have convinced myself of the latter and agree with you. We get to choose what is normal and supported behaviour and I can't see a downside to supporting the deletion of jobs.

I'll update this to introduce a destroyed phase which will be set if the job is missing with the same behaviour as the stopped phase - requeue after the TTL and and GC the console.

File: pkg/workloads/console/acceptance/acceptance.go [hunk 2]
```diff
@@ -175,6 +175,56 @@ func (r *Runner) Run(logger kitlog.Logger, config *rest.Config) {
...
+			err = client.BatchV1().Jobs(namespace).Delete(jobName, &metav1.DeleteOptions{PropagationPolicy: &policy})
+			Expect(err).NotTo(HaveOccurred(), "could not delete console job")
+
+			By("Expect that the job no longer exists")
+			Eventually(func() error {
+				_, err = client.BatchV1().Jobs(namespace).Get(consoleName, metav1.GetOptions{})
+				return err
+			}).Should(HaveOccurred(), "expected not to find job, but did")
+
+			time.Sleep(time.Second)
```
Comment (benwh) [inline pkg/workloads/console/acceptance/acceptance.go:219]: This seems a bit racy. Could we instead do a check that *eventually* the console's status changes to stopped (or destroyed)?

Comment (dyson) [reply pkg/workloads/console/acceptance/acceptance.go:219]: Yeah this is definitely racey and we can now check with eventually that its status is updated to destroyed like you mention.

**Comment by dyson** [comment]
This needs a new full review as lots has changed since the original PR. 

We now mark consoles as destroyed if the job has been deleted and perform GC on them the same as if the console had been stopped.

Also while doing this work I noticed we watch on all console events. This meant that when we updated a consoles status, we would trigger another reconciliation. I have changed this to only watch for console creations. All other reconciliations should be triggered by events happening on the child objects we watch.

Finally, in allowing for jobs to be deleted, more parts of the code needed to be changed compared to the original implementation as were no longer exiting reconciliation early with an error -  reconciliation now needs to fully support a console with no job.

File: pkg/workloads/console/controller.go [hunk 3]
```diff
@@ -347,21 +383,30 @@ func (r *reconciler) clampTimeout(template *workloadsv1alpha1.ConsoleTemplate) e
...
-		InNamespace(r.name.Namespace).
-		MatchingLabels(map[string]string{"job-name": job.ObjectMeta.Name})
-	if err := r.client.List(r.ctx, opts, podList); err != nil {
-		return err
-	}
-	if len(podList.Items) > 0 {
-		pod = &podList.Items[0]
+	var newStatus workloadsv1alpha1.ConsoleStatus
+
+	if job == nil {
```
Comment (benwh) [inline pkg/workloads/console/controller.go:388]: Putting a conditional here seems a bit wrong. Can we not instead conditionally wrap the bit of the code that does the pod listing so that it doesn't attempt to run if the job is nil.

Then we can keep doing all of the status calculation stuff in the `calculateStatus` func?

File: pkg/workloads/console/controller.go [hunk 4]
```diff
@@ -388,6 +433,10 @@ func (r *reconciler) updateStatus(job *batchv1.Job) error {
 		auditLog(r.logger, r.console, &duration, ConsoleEnded, "Console ended after reach expiration time")
 	}
 
+	if newStatus.Phase == workloadsv1alpha1.ConsoleDestroyed {
```
Comment (benwh) [inline pkg/workloads/console/controller.go:436]: The fact that we're always logging when the phase is destroyed, rather than only when we transition to destroyed, seems odd? Is this related to why we've got the predicates added?

File: pkg/workloads/console/controller.go (hunk 3 above)
Comment (benwh) [reply pkg/workloads/console/controller.go:388]: Yeah after looking at the changes in the `calculateStatus` func below, I do think it'd be nicer to keep this logic in the `calculatePhase` function

File: pkg/workloads/console/controller.go [hunk 5]
```diff
@@ -102,7 +117,9 @@ func Add(ctx context.Context, logger kitlog.Logger, mgr manager.Manager, opts ..
 	}
 
 	err = ctrl.Watch(
-		&source.Kind{Type: &workloadsv1alpha1.Console{}}, &handler.EnqueueRequestForObject{},
+		&source.Kind{Type: &workloadsv1alpha1.Console{}},
+		&handler.EnqueueRequestForObject{},
+		CreatePredicate{},
```
[the remark this answers is not in the document]
Comment (dyson) [reply pkg/workloads/console/controller.go:122]: Spoke about this and going to revert the change. Essentially calling the extra reconciles on update are harmless and we want to support manual updates as we could, although in practice we wouldn't, edit a console. If we don't listen to update events, our console could end up out os sync.

File: pkg/workloads/console/controller.go [hunk 6]
```diff
@@ -182,10 +199,20 @@ func (r *reconciler) Reconcile() (res reconcile.Result, err error) {
...
-		return res, err
+	var job *batchv1.Job
+
+	// Try to get the consoles job
+	if j, err := r.getJob(); err == nil {
+		job = j
+	}
+
+	// Only create/update a job when the console is creating or when one
+	// already exists
```
[the remark this answers is not in the document]
Comment (dyson) [reply pkg/workloads/console/controller.go:210]: Yep! Much better!

File: pkg/workloads/console/controller.go [hunk 7]
```diff
@@ -213,7 +239,7 @@ func (r *reconciler) Reconcile() (res reconcile.Result, err error) {
 			return res, err
 		}
 
-	case r.console.Stopped():
+	case r.console.Stopped() || r.console.Destroyed():
```
[the remark this answers is not in the document]
Comment (dyson) [reply pkg/workloads/console/controller.go:244]: Agreed on adding an `Active()` helper and using it in the two places.

File: pkg/workloads/console/controller.go (hunk 3 above)
Comment (dyson) [reply pkg/workloads/console/controller.go:390]: Agreed about this. It was wrong to set a phase outside of `calculatePhase`. I'll rejig this so we do it in there.

File: pkg/workloads/console/controller.go (hunk 4 above)
Comment (dyson) [reply pkg/workloads/console/controller.go:438]: Ah yeah good catch! I'll change the conditional to only log when the existing state isn't already `workloadsv1alpha1.ConsoleDestroyed`.
